src/tzdb.rs

```diff
@@ -329,7 +375,117 @@ fn get_timezone_offset(db: &DataBlock, idx: usize) -> TimeZoneTransitionInfo {
+    debug_assert!(
+        transition_time.is_some(),
+        "tzif internal invariant violated"
+    );
+    let transition_time = transition_time.copied().unwrap_or_default();
+    let prev = if idx == 0 {
+        db.local_time_type_records[0]
+    } else {
+        get_local_record(db, idx - 1)
+    };
+    TzifTransitionInfo {
+        prev,
+        next,
+        transition_time,
+    }
+}
+
+/// Information obtained from the tzif file about a transition.
+#[derive(Debug)]
+struct TzifTransitionInfo {
+    /// The time record from before this transition
+    prev: LocalTimeTypeRecord,
+    /// The time record corresponding to this transition and dates after it
+    next: LocalTimeTypeRecord,
+    /// The UTC epoch seconds for the transition
+    transition_time: Seconds,
+}
+
+impl TzifTransitionInfo {
+    /// Get the previous offset as a number of seconds from
+    /// 1970-01-01 in local time as reckoned by the previous offset
+    fn transition_time_prev_epoch(&self) -> Seconds {
+        // You always add the UTC offset to get the local time;
+        // so a local time in PST (-08:00) will be `utc - 8h`
+        self.transition_time + self.prev.utoff
+    }
+    /// Get the previous offset as a number of seconds from
+    /// 1970-01-01 in local time as reckoned by the next offset
+    fn transition_time_next_epoch(&self) -> Seconds {
+        // You always add the UTC offset to get the local time;
+        // so a local time in PST (-08:00) will be `utc - 8h`
+        self.transition_time + self.next.utoff
+    }
+
+    /// Gets the range of local times where this transition is active
+    ///
+    /// Note that this will always be start..end, NOT prev..next: if the next
+    /// offset is before prev (e.g. for a TransitionKind::Overlap) year,
+    /// it will be next..prev.
+    ///
+    /// You should use .kind() to understand how to interpret this
+    fn offset_range_local(&self) -> Range<Seconds> {
+        let prev = self.transition_time_prev_epoch();
+        let next = self.transition_time_next_epoch();
+        match self.kind() {
+            TransitionKind::Overlap => next..prev,
+            _ => prev..next,
+        }
+    }
+
+    /// What is the kind of the transition?
+    fn kind(&self) -> TransitionKind {
+        match self.prev.utoff.cmp(&self.next.utoff) {
+            Ordering::Less => TransitionKind::Gap,
+            Ordering::Greater => TransitionKind::Overlap,
+            Ordering::Equal => TransitionKind::Smooth,
+        }
+    }
+
+    /// If a time is found to be within self.offset_range_local(),
+    /// what is the corresponding LocalTimeRecordResult?
+    fn record_for_contains(&self) -> LocalTimeRecordResult {
+        match self.kind() {
+            TransitionKind::Gap => LocalTimeRecordResult::Empty,
+            TransitionKind::Overlap => {
+                if self.prev.is_dst {
+                    return LocalTimeRecordResult::Ambiguous {
+                        dst: self.prev.into(),
```

Comment:
this isn't correct.

This type should just have fields first and second, not DST and standard, since it ends up turning into a list, which is expected to be sorted.

Reply:
I tried fixing this and started failing more tests. This is trickier than previously thought.